src/openlifu/plan/solution.py

```diff
@@ -1,7 +1,169 @@
-from dataclasses import dataclass
+import base64
+import json
+from dataclasses import asdict, dataclass, field
+from datetime import datetime
+from pathlib import Path
+from typing import Optional
 
+import numpy as np
+import xarray
+
+from openlifu.bf import Pulse, Sequence
+from openlifu.geo import Point
+from openlifu.util.json import PYFUSEncoder
+
+
+def _construct_nc_filepath_from_json_filepath(json_filepath:Path) -> Path:
+    """Construct a default filepath to netCDF file given filepath to associated solution json file."""
+    nc_filename = json_filepath.name.split(".")[0] + ".nc"
+    nc_filepath = json_filepath.parent / nc_filename
+    return nc_filepath
 
 @dataclass
 class Solution:
-    id: str = "solution"
+    """
+    A sonication solution resulting from beamforming and running a simulation.
+    """
+    id: str = "solution" # the *solution* id, a concept that did not exist in the matlab software
+    """ID of this solution"""
+
     name: str = "Solution"
+    """Name of this solution"""
+
+    protocol_id: Optional[str] = None # this used to be called plan_id in the matlab code
+    """ID of the protocol that was used when generating this solution"""
+
+    transducer_id: Optional[str] = None
+    """ID of the transducer that was used when generating this solution"""
+
+    created_on: datetime = field(default_factory=datetime.now)
+    """Solution creation time"""
+
+    description: str = ""
+    """Description of this solution"""
+
+    delays: Optional[np.ndarray] = None
+    """Vector of time delays to steer the beam"""
+
+    apodizations: Optional[np.ndarray] = None
+    """Vector of apodizations to steer the beam"""
+    pulse: Pulse = field(default_factory=Pulse)
+    """Pulse to send to the transducer when running sonication"""
+    sequence: Sequence = field(default_factory=Sequence)
+    """Pulse sequence to use when running sonication"""
+    focus: Optional[Point] = None
+    """Point that is being focused on in this Solution; part of the focal pattern of the target"""
+
+    # there was "target_id" in the matlab software, but here we do not have the concept of a target ID.
+    # I believe this was only needed in the matlab software because solutions were organized by target rather
+    # than having their own unique solution ID. We do have unique solution IDs so it's possible we don't need
+    # this target attribute at all here. Keeping it here for now just in case.
+    target: Optional[Point] = None
+    """The ultimate target of this sonication. This sonication solution is focused on one focal point
+    in a pattern that is centered on this target."""
+
+    # In the matlab code the simulation result was saved as a separate .mat file.
+    # Here we include it as an xarray dataset.
+    simulation_result: xarray.Dataset = field(default_factory=xarray.Dataset)
+    """The xarray Dataset of simulation results"""
```

Comment:
Are you talking about what was previously `solutions/example_plan/example_target.mat` ?

Reply:
I believe s. The `simulation_result` would contain the `ds` object [returned by `run_simulation`](https://github.com/OpenwaterHealth/OpenLIFU-python/blob/68166a29cbe564630b7c9933f36692060a1bb509/src/openlifu/sim/kwave_if.py#L127)

So this would be an xarray Dataset containing "pnp" (peak negative pressure), "ppp" (peak positive pressure), and "ita" (time average intensity) arrays defined over the simulation grid.